### PR TITLE
tests: loader's tests for errors and warnings

### LINKED ISSUE #1581: webpack-cli v4-beta.8 ignoring loader's error.
**Describe the bug**

webpack-cli v4-beta.8 ignoring loader's error.

Reproduction code: https://github.com/mstssk/webpack-cli-v4-beta8-ignore-loader-error

I wrote reproduction code of ts-loader, but this problem occurs with other loaders. e.g. file-loader.

**What is the current behavior?**

webpack-cli v3.3.11 fails and returns exit code 2 when an error occurs in loaders.
But webpack-cli v4.0.0-beta.8 does not fail and return exit code 0.

**To Reproduce**

1. Setup webpack@4.43.0, webpack-cli@4.0.0-beta.8, and some loaders
2. Write codes which raise error in loaders.
3. Run webpack
4. See error logs and status code.

**Expected behavior**

webpack fails and returns non-zero exit code when an error occurs in loaders.

**Screenshots**

**Please paste the results of `webpack-cli info` here, and mention other relevant information**

`webpack-cli info` fails in my env.

<details>
<summary>`webpack-cli info` Log</summary>

```
 $ npx webpack-cli info
[webpack-cli] Promise rejection: TypeError: Cannot convert undefined or null to object
[webpack-cli] TypeError: Cannot convert undefined or null to object
    at Function.entries (<anonymous>)
    at Object.<anonymous> (/private/tmp/webpack-cli-v4-beta8-ignore-loader-error/node_modules/@webpack-cli/info/index.js:45:41)
    at Generator.next (<anonymous>)
    at /private/tmp/webpack-cli-v4-beta8-ignore-loader-error/node_modules/@webpack-cli/info/index.js:8:71
    at new Promise (<anonymous>)
    at __awaiter (/private/tmp/webpack-cli-v4-beta8-ignore-loader-error/node_modules/@webpack-cli/info/index.js:4:12)
    at Object.info [as default] (/private/tmp/webpack-cli-v4-beta8-ignore-loader-error/node_modules/@webpack-cli/info/index.js:44:12)
    at Function.run (/private/tmp/webpack-cli-v4-beta8-ignore-loader-error/node_modules/webpack-cli/lib/commands/ExternalCommand.js:34:51)
    at WebpackCLI.runCommand (/private/tmp/webpack-cli-v4-beta8-ignore-loader-error/node_modules/webpack-cli/lib/webpack-cli.js:286:60)
    at runCLI (/private/tmp/webpack-cli-v4-beta8-ignore-loader-error/node_modules/webpack-cli/lib/bootstrap.js:61:26)
```

</details>

**Additional context**

https://github.com/webpack/webpack-cli/issues/1222#issuecomment-633485026

### PR DESCRIPTION
**Pull request by snitin315**
**What kind of change does this PR introduce?**
test

**Did you add tests for your changes?**
Yes
**If relevant, did you update the documentation?**
NA
**Summary**
Fixes #1581 

**Does this PR introduce a breaking change?**
No

**Other information**

### CODE REVIEW

File: test/loader/error-test/loader-error.test.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
+'use strict';
+
+// eslint-disable-next-line node/no-unpublished-require
+const { run } = require('../../utils/test-utils');
+
+describe('loader error regression test for #1581', () => {
+    it(`should not ignore loader's error produce a failing build`, () => {
+        const { stderr, stdout } = run(__dirname, [], false);
+        expect(stderr).toBeFalsy();
+        expect(stdout).toContain('[built] [failed] [1 error]');
```
Comment (snitin315) [inline test/loader/error-test/loader-error.test.js:10]:
```suggestion
        expect(stdout).toContain('[1 error]');
```

**Comment by evilebottnawi** [comment]
deprecations from ts-loader?

**Comment by snitin315** [comment]
Yes, can we test on `stdout` only?

File: test/loader/error-test/loader-error.test.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
+'use strict';
+
+// eslint-disable-next-line node/no-unpublished-require
+const { run } = require('../../utils/test-utils');
+
+describe('loader error regression test for #1581', () => {
+    it(`should not ignore loader's error produce a failing build`, () => {
+        const { stderr, stdout } = run(__dirname, [], false);
+        expect(stderr).toBeFalsy();
```
Comment (snitin315) [inline test/loader/error-test/loader-error.test.js:9]:
```suggestion
        const { stdout } = run(__dirname, [], false);
```

**Comment by evilebottnawi** [comment]
Yes, let comment `stderr` (with comment on issue), open an issue in `ts-loader` and merge it

File: test/loader/error-test/loader-error.test.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+const { run } = require('../../utils/test-utils');
+
+describe('loader error regression test for #1581', () => {
+    it(`should not ignore loader's error produce a failing build`, () => {
+        // Ignoring assertion on stderr because ts-loader is producing depreciation warnings
+        // with webpack@v5.0.0-beta.24 -> https://github.com/TypeStrong/ts-loader/issues/1169
+        const { stdout } = run(__dirname, [], false);
+
+        expect(stdout).toContain('[1 error]');
+        expect(stdout).toContain(`Cannot assign to 'foobar' because it is a constant`);
```
Comment (evilebottnawi) [inline test/loader/error-test/loader-error.test.js:13]: Can we add test on errorCode?

Comment (snitin315) [reply test/loader/error-test/loader-error.test.js:13]: Any suggestion on how can we drive webpack's error code here?

Comment (evilebottnawi) [reply test/loader/error-test/loader-error.test.js:13]: execa should exit with same code, just check `code` from execa, it should be not `0`

Comment (snitin315) [reply test/loader/error-test/loader-error.test.js:13]: It is `0` 😞 .

Comment (evilebottnawi) [reply test/loader/error-test/loader-error.test.js:13]: It is bug, I think it https://github.com/webpack/webpack-cli/pull/1739 should fix it, so our process will have original exit code

Comment (evilebottnawi) [reply test/loader/error-test/loader-error.test.js:13]: Maybe you can help @anshumanv with this PR

**Comment by snitin315** [comment]
still same 😞

**Comment by evilebottnawi** [comment]
@snitin315 We need to solve it before release, there are two possible problems:
- our test utils is not respect exit code
- something wrong inside `webpack-cli` logic in `try/catch`

**Comment by snitin315** [comment]
Yes, I'm looking into it.

File: packages/webpack-cli/lib/utils/Compiler.js [hunk 4]
```diff
@@ -93,9 +96,6 @@ class Compiler {
             }
             return this.generateOutput(outputOptions, stats, statsErrors, processingMessageBuffer);
         }
-        if (!outputOptions.watch && stats.hasErrors()) {
-            process.exitCode = 2;
-        }
     }
```
Comment (snitin315) [inline packages/webpack-cli/lib/utils/Compiler.js:99]: This block was never executed because we have a return statement earlier.

Comment (evilebottnawi) [reply packages/webpack-cli/lib/utils/Compiler.js:99]: We should refactor it and avoid using `return`

File: packages/webpack-cli/lib/utils/Compiler.js [hunk 5]
```diff
@@ -80,6 +80,9 @@ class Compiler {
             logger.error(err.stack || err);
             process.exit(1); // eslint-disable-line
         }
+        if (!outputOptions.watch && stats.hasErrors()) {
+            process.exitCode = 1;
+        }
```
Comment (evilebottnawi) [inline packages/webpack-cli/lib/utils/Compiler.js:85]: warnings and errors with `1`, here only errors, we need to test on warnings too, you can create simple loader and use `this.emitWarning` to test it

File: test/loader/warning-test/src/main.js [hunk 6]
```diff
@@ -0,0 +1 @@
+console.log('loader warning test');
```
Comment (evilebottnawi) [inline test/loader/warning-test/src/main.js:1]: we need to use this loader in `main.js`

**Review by evenstensberg — APPROVED** [review]

**Review by evilebottnawi — APPROVED** [review]

**Review by anshumanv — APPROVED** [review]

**Review by jamesgeorge007 — APPROVED** [review]

**Pull request merged**